src/net_packet.c

```diff
@@ -193,15 +193,7 @@ static void udp_probe_h(node_t *n, vpn_packet_t *packet, length_t len) {
 	/* It's a valid reply: now we know bidirectional communication
 	   is possible using the address and socket that the reply
 	   packet used. */
-	if(!n->status.udp_confirmed) {
-		n->status.udp_confirmed = true;
-
-		if(!n->address_cache) {
-			n->address_cache = open_address_cache(n);
-		}
-
-		reset_address_cache(n->address_cache, &n->address);
-	}
```

Comment:
But sometimes we only learn the correct address via UDP. For example, nodes A, B and C, A connects to B, B connects to C, but B doesn't have the correct address of A (maybe due to NAT somewhere), but A learns the correct address from C, then A sends C UDP packets, without NAT being in the way, and without initiating a meta-connection (because the autoconnect algorithm might already be satisfied somehow).

Reply:
Sure, but `n->address` is a UDP address, and it means possibly pushing out cached TCP addresses. `n->connection->address` is not guaranteed to be initialized here.

I'm testing all this on a 15-node network with multiple firewalls and layers of NAT in front of some nodes.

If this change is reverted to the original state, tincd starts replacing perfectly working TCP ports that were cached in other places with other ports that look weird (like 772) and don't accept incoming connections. After restarting tincd you get a bunch of "Connection refused" from nodes that could be connected to before.

I'm probably missing something obvious here. I'll keep this change for now as it works better without this code, and dig more into tinc internals. Everything else should be fixed.

Reply:
But we should know the TCP port `n` is listening on, as that is in the edge information. But yes, we can keep this change and revisit this in another ticket.

Reply:
Right, the edge address always seems to be valid in my testing.